### PR TITLE
Improve post test sanity check

### PR DESCRIPTION
**Pull request by wangxin**
### Description of PR

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [x] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

#### What is the motivation for this PR?
Currently if post test sanity check is enabled, the exactly same set of checks
will be performed just like pre test sanity check. Sometimes we may want to
perform different set of checks for post test sanity check.

#### How did you do it?
This change improved sanity check in many ways:
* Added --post_check command line option.
* Added --post_check_items command line option for fine tuning post test sanity
  check items.
* Added 'post_check_items' keyword argument for the sanity_check marker for fine
  tuning post test sanity check items.
* Added support of passing arguments to the check functions defined in check
  fixtures.
* Added argument 'stage' to 'do_checks' function. Then the check functions
  would be able to know the stage (pre or post test) of checking.
* Used shell_cmds module to run the batch of commands for collecting logs in
  the print_logs function. This change reduced test execution time.
* Updated the document.

#### How did you verify/test it?
Tried different values for the sanity check related options in both sanity_check marker and command line.

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -93,6 +93,10 @@ def pytest_addoption(parser):
                      help="Allow recovery attempt in sanity check in case of failure")
     parser.addoption("--check_items", action="store", default=False,
                      help="Change (add|remove) check items in the check list")
+    parser.addoption("--post_check", action="store_true", default=False,
+                     help="Perform post test sanity check if sanity check is enabled")
+    parser.addoption("--post_check_items", action="store", default=False,
```
Comment (lolyu) [inline tests/conftest.py:98]: So now with your change, the `post_check` is default off, right?

File: tests/common/plugins/sanity_check/__init__.py [hunk 2]
```diff
@@ -79,18 +79,36 @@ def _update_check_items(old_items, new_items, supported_items):
...
+
+    return filtered_check_items
+
+
+def do_checks(request, check_items, *args, **kwargs):
     check_results = []
     for item in check_items:
         check_fixture = request.getfixturevalue(_item2fixture(item))
-        results = check_fixture()
+        results = check_fixture(*args, **kwargs)
```
Comment (xumia) [inline tests/common/plugins/sanity_check/__init__.py:111]: Hi Xin, can we add the request in parameters? So we have the context info in the sanity check, so we can save the info per module name.

File: tests/conftest.py (hunk 1 above)
Comment (wangxin) [reply tests/conftest.py:98]: Both before and with this change, `post_check` is default off. The default behavior is not changed.

File: tests/common/plugins/sanity_check/__init__.py (hunk 2 above)
Comment (wangxin) [reply tests/common/plugins/sanity_check/__init__.py:111]: When you define a new check, the check must be a fixture that returns a function that is responsible for the actual checking. The `request` object itself is a pytest fixture. So, you can get the `request` object just in your check fixture by specifying it in the argument list. For example:
```
@pytest.fixture(scope='module')
def check_something(request):
    def _check(*args, **kwargs):
        # Run checking
        # The request object is available in this scope.
        return check_result
    return _check
```

**Review by xumia — APPROVED** [review]

**Pull request merged**